### ISSUE TITLE
Cannot install Rook using DO PVs -  /var/lib/kubelet/plugins/dobs.csi.digitalocean.com/csi.sock: connect: connection refused

### ISSUE THREAD

**Issue by MasterNayru** [digitalocean/csi-digitalocean#209]
### What did you do? (required. The issue will be **closed** when not provided.)
I am trying to install Rook on my Kubernetes cluster using DigitalOcean PVCs as the backing storage/mon disks. The manifests and setup instructions I am using are pretty much straight from the getting started docs on Rook's site, but it seems that Rook is having its connections to the DigitalOcean CSI Unix socket blocked with the following events on the Rook OSD prepare pods:
```
Events:
  Type     Reason           Age                    From                  Message
  ----     ------           ----                   ----                  -------
  Warning  FailedMount      12m (x520 over 19h)    kubelet, k8sworker02  (combined from similar events): Unable to attach or mount volumes: unmounted volumes=[set1-0-data-fch6l], unattached volumes=[rook-data udev rook-binaries rook-ceph-osd-token-wjd6x set1-0-data-fch6l-bridge set1-0-data-fch6l ceph-conf-emptydir rook-ceph-log devices]: timed out waiting for the condition
  Warning  FailedMapVolume  5m58s (x399 over 20h)  kubelet, k8sworker02  MapVolume.SetUp failed for volume "pvc-2578a0e1-ba2d-4144-86a2-c2d03ad73642" : kubernetes.io/csi: blockMapper.stageVolumeForBlock failed to check STAGE_UNSTAGE_VOLUME capability: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial unix /var/lib/kubelet/plugins/dobs.csi.digitalocean.com/csi.sock: connect: connection refused"
  Warning  FailedMapVolume  114s (x203 over 20h)   kubelet, k8sworker02  MapVolume.SetUp failed for volume "pvc-2578a0e1-ba2d-4144-86a2-c2d03ad73642" : kubernetes.io/csi: blockMapper.stageVolumeForBlock failed: rpc error: code = Unavailable desc = transport is closing
```

The volumes themselves have been created, and bound to PVCs correctly as shown here, so I assume that I have set up the driver correctly:
```
 ~ $ kubectl get pvc -n rook-ceph     
NAME                STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS       AGE
rook-ceph-mon-a     Bound    pvc-5d2f3a71-cccb-46ba-9ef1-32b1d04046a6   10Gi       RWO            do-block-storage   20h
rook-ceph-mon-b     Bound    pvc-6c503d86-713b-475c-b5f7-e2c8b7fb16d8   10Gi       RWO            do-block-storage   20h
rook-ceph-mon-c     Bound    pvc-bb7fac52-1888-49ff-a067-e1e5f8f5cbde   10Gi       RWO            do-block-storage   20h
set1-0-data-fch6l   Bound    pvc-2578a0e1-ba2d-4144-86a2-c2d03ad73642   50Gi       RWO            do-block-storage   20h
set1-1-data-l9h6j   Bound    pvc-b071e169-272d-4f39-ad9b-d5bbb4e15d95   50Gi       RWO            do-block-storage   20h
set1-2-data-mcdz2   Bound    pvc-fde4652b-b94d-41ea-a4ec-c2d476afe2ff   50Gi       RWO            do-block-storage   20h
```
It would appear that carrying out the mount of the voulmes on the host requires communication with the above socket, and those connections are being refused for some reason.

### What did you expect to happen?
The volumes, having been created successfully and bound to PVCs, would be able to be mounted on the host and Rook would be able to do its thing. Without the ability to do this, I am struggling to understand how to minimise downtime on stateful Kubernetes deployments on DO since Rook is, to my knowledge, the best practice solution for having highly available storage in Kubernetes.

### Configuration (**MUST** fill this out):

* system logs:
I don't want to dump my whole cluster info into this ticket, but I was able to reproduce this pretty easily with the following manifest: https://gist.github.com/MasterNayru/b7186eae4c68561e33b0d71620a24db8

* CSI Version: v1.1.2 (and whatever is installed in DOKS clusters)

* Kubernetes Version: Have reproduced this on both 1.15.5 and 1.16.2

* Cloud provider/framework version, if applicable (such as Rancher):
I have tried both DOKS v1.15.5 and v1.16.2. After sending a message to DO Support, they suggested that this may be due to an implementation detail, so I spun up my own Kubernetes v1.16.2 cluster and installed the CSI v1.1.2 release, popped my token in a secret, applied the above manifests and got to the same error.

**Comment by timoreimann** [comment]
hey @MasterNayru 👋 

I believe the problem is that [Rook requires block volume mode](https://gist.github.com/MasterNayru/b7186eae4c68561e33b0d71620a24db8#file-cluster-on-pvc-digitalocean-yaml-L75). This is not something our CSI driver provides today, but we can implement support for it now that the CSI spec supports it. Issue #192 is tracking the effort, and it's probably the next thing we're going to tackle (unless someone from the community beats us to it).

Let's keep this issue open for visibility and double-checking purposes since Rook seems quite popular.

Thanks.

**Comment by zetas** [comment]
+1 for this. I've been tearing my hair out the last few days trying to debug this issue and get rook + volume osd's running properly.

**Comment by adamwg** [comment]
@MasterNayru @zetas If you wouldn't mind sharing, what's your use-case for Rook/ceph on top of DO block storage volumes? I ask because DO block storage is already Ceph on the backend - volumes are 3-way replicated and highly available.

To be clear, I'm not questioning whether we should support running Rook with volumes on DOKS - we should definitely implement block volume mode and support it. Just curious what folks are getting from it beyond what our block storage volumes already offer.

**Comment by zetas** [comment]
It's funny you should say that, my brother is a big data sales engineer at redhat and when I told him what I was doing he was like "DigitalOcean already uses ceph for block storage, why are you putting ceph on top of ceph?" and I was like "lol because money". Here's the main reasons:

1. I already use my k8s cluster for other stuff, (openfass, mysql, etc), I would love to be able to use all that drive space that usually sits wasted for scratch storage for my other droplets. 

2. Not being able to mount/attach volumes to multiple droplets is a big issue as well. CephFS allows me to have a shared filesystem that I can consume on many external droplets. I was limited to NFS before and we all know performance for NFS is garbage. 

3. As far as using your block storage volumes as OSD's my idea was to be able to extend the ceph storage incrementally, spinning up smaller DO volumes, 10 to 50gb at a time and adding them or removing them as needed via PVC's. It's much faster (and cheaper) than spinning up a new k8 node for added storage. I could even do this automatically which is a huge attraction to me as a devops engineer that likes to automate all the things. 

## In conclusion 
Your volumes are great but at $10 per 100gb it's far too expensive for what I need it for. I need essentially an elastic volume that can scale up to 700-1000gb and then go back down but I dont want to pay $70-$100 a month for the rare opportunity when I might need that much space due to a large processing event on my droplets (and if the droplets run out of space they just kinda hang there and never progress until I intervene, not ideal)

I know that using ceph for this is crazy overkill; I actually generally disable replication on the pools or at least set it to the minimum amount of 2 to maximize the amount of space available. I just really like the idea of being able to leverage every resource I pay for. When you're running a small business every dollar saved counts.

**Comment by adamwg** [comment]
@zetas Thanks for the detailed reply! Sounds like an interesting and totally reasonable set of use-cases. The idea of using the local storage and then "bursting" to block is especially interesting. I wonder whether we could do something on our side to make that easier - something I'll have to think about.

**Comment by timoreimann** [comment]
We merged support for raw block mode into master, and rudimentary validated it works with Rook.

The next CSI and DOKS releases are going to ship the feature.

**Comment by MasterNayru** [comment]
Woohoo! Thanks, everyone! Greatly appreciated. Eagerly awaiting the next release to give this a go

**Comment by verenion** [comment]
Is this issue fixed, as I'm still seeing this issue when my three `rook-ceph-osd-prepare-set` is trying to spin up.

**Comment by timoreimann** [comment]
@verenion are you using the latest CSI driver v1.2.0? If you're a DOKS user, you need to make sure that you're on a [version supporting raw block mode](https://www.digitalocean.com/docs/kubernetes/resources/volume-features/).

**Comment by h0jeZvgoxFepBQ2C** [comment]
If ceph is running anyway in the digital ocean backend, why the h*** don't you offer a ceph-fs backend so we can ReadWriteMany easily :) ? We are trying to get rook-ceph running on DO, just to circumvent this restriction (and single node NFS reliability was really poor during node recycling etc). The missing ReadWriteMany access more is really a big minus on DO.. would be great if you could offer this?
